### PR TITLE
cosmo: init at 0.18.4

### PR DESCRIPTION
**Pull request by rvolosatovs**
## Description of changes

Add `cosmo`, the CLI for https://cosmonic.com/, which is based on open-source https://github.com/wasmCloud/wasmcloud CNCF sandbox project

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by rvolosatovs** [comment]
@ofborg build cosmo

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 257247` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cosmo</li>
  </ul>
</details>

File: pkgs/development/tools/cosmo/default.nix [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        hash
+        ;
+      url = "https://cosmonic.sh/cosmo-${system}-${version}";
+    };
+
+    installPhase = ''
+      mkdir -p $out/bin
+      cp $src $out/bin/cosmo
+      chmod +x $out/bin/cosmo
+    '';
```
Comment (ocfox) [inline pkgs/development/tools/cosmo/default.nix:40]: Don't forget the runHook.

**Review by ocfox — CHANGES_REQUESTED** [review]
Other than the weird format and lengthy description, LGTM!

**Comment by ocfox** [comment]
Result of `nixpkgs-review pr 257247` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cosmo</li>
  </ul>
</details>

**Comment by rvolosatovs** [comment]
> Other than the weird format and lengthy description, LGTM!

What do you mean by the "weird format"?

**Comment by ocfox** [comment]
> What do you mean by the "weird format"?

The code format is not quite in the form normally used by nixpkgs.

**Comment by rvolosatovs** [comment]
Rebased to address a merge conflict, could someone take a look?

File: pkgs/development/tools/cosmo/default.nix [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+      runHook postInstall
+    '';
+
+    meta = with lib; {
+      description = "Leverage the power and flexibility of WebAssembly. Take your ideas from concept to deployed anywhere, at scale, in minutes.";
+      homepage = "https://cosmonic.com/";
+      license = licenses.unfree;
+      maintainers = with maintainers; [rvolosatovs];
+      platforms = with platforms; linux ++ darwin ++ windows;
```
Comment (pbsds) [inline pkgs/development/tools/cosmo/default.nix:51]: platforms has a lot more systems strings than what you have hashes for

File: pkgs/development/tools/cosmo/default.nix [hunk 3]
```diff
@@ -0,0 +1,53 @@
+{
+  lib,
+  stdenvNoCC,
+  fetchurl,
+}: let
```
Comment (pbsds) [inline pkgs/development/tools/cosmo/default.nix:5]:
```suggestion
}: 
let
```

File: pkgs/development/tools/cosmo/default.nix [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      aarch64-darwin = "sha256-Nk/mhtgdOISc9JbnZAEQ/g6ubie+SnRM729zpG6/lo4=";
+      aarch64-linux = "sha256-lPXbbF/Zr1+XRB04KxnVt7DxpJCN5xCinTxPDDg4h2Y=";
+      x86_64-darwin = "sha256-tCjhQXsfyePHHeTz3zoz+dp9MPGo7ST1bGQjaM1OBgk=";
+      x86_64-linux = "sha256-bwvOJAr0KoAI808OaGZB5uhG1J0dzrodr3ZXK5/9S7Q=";
+      x86_64-windows = "sha256-3W/QzY7jcVxGkHsEsfNyhvDzca2TC/sTxN5/AOFZz/g=";
+    }
+    .${system}
+    or (throw "Unsupported system: ${system}");
+in
+  stdenvNoCC.mkDerivation rec {
```
Comment (pbsds) [inline pkgs/development/tools/cosmo/default.nix:20]:
```suggestion
in
stdenvNoCC.mkDerivation rec {
```

and dedent the rest

File: pkgs/development/tools/cosmo/default.nix [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  hash =
+    {
+      aarch64-darwin = "sha256-Nk/mhtgdOISc9JbnZAEQ/g6ubie+SnRM729zpG6/lo4=";
+      aarch64-linux = "sha256-lPXbbF/Zr1+XRB04KxnVt7DxpJCN5xCinTxPDDg4h2Y=";
+      x86_64-darwin = "sha256-tCjhQXsfyePHHeTz3zoz+dp9MPGo7ST1bGQjaM1OBgk=";
+      x86_64-linux = "sha256-bwvOJAr0KoAI808OaGZB5uhG1J0dzrodr3ZXK5/9S7Q=";
+      x86_64-windows = "sha256-3W/QzY7jcVxGkHsEsfNyhvDzca2TC/sTxN5/AOFZz/g=";
+    }
+    .${system}
+    or (throw "Unsupported system: ${system}");
```
Comment (pbsds) [inline pkgs/development/tools/cosmo/default.nix:18]:
```suggestion
  hashes = {
    aarch64-darwin = "sha256-Nk/mhtgdOISc9JbnZAEQ/g6ubie+SnRM729zpG6/lo4=";
    aarch64-linux = "sha256-lPXbbF/Zr1+XRB04KxnVt7DxpJCN5xCinTxPDDg4h2Y=";
    x86_64-darwin = "sha256-tCjhQXsfyePHHeTz3zoz+dp9MPGo7ST1bGQjaM1OBgk=";
    x86_64-linux = "sha256-bwvOJAr0KoAI808OaGZB5uhG1J0dzrodr3ZXK5/9S7Q=";
    x86_64-windows = "sha256-3W/QzY7jcVxGkHsEsfNyhvDzca2TC/sTxN5/AOFZz/g=";
  }
```

File: pkgs/development/tools/cosmo/default.nix [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      ;
+    pname = "cosmo";
+
+    preferLocalBuild = true;
+    dontUnpack = true;
+
+    src = fetchurl {
+      inherit
+        hash
+        ;
```
Comment (pbsds) [inline pkgs/development/tools/cosmo/default.nix:32]:
```suggestion
      hash =  hashes.${system} or (throw "Unsupported system: ${system}");
```

File: pkgs/development/tools/cosmo/default.nix (hunk 2 above)
Comment (pbsds) [reply pkgs/development/tools/cosmo/default.nix:51]:
```suggestion
      platforms = lib.attrNames hashes;
```

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -20890,6 +20890,8 @@ with pkgs;
 
   cog = callPackage ../development/web/cog { };
 
+  cosmo = callPackage ../development/tools/cosmo { };
```
Comment (pbsds) [inline pkgs/top-level/all-packages.nix:20893]: please migrate to by-name

File: pkgs/development/tools/cosmo/default.nix [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      x86_64-linux = "sha256-bwvOJAr0KoAI808OaGZB5uhG1J0dzrodr3ZXK5/9S7Q=";
+      x86_64-windows = "sha256-3W/QzY7jcVxGkHsEsfNyhvDzca2TC/sTxN5/AOFZz/g=";
+    }
+    .${system}
+    or (throw "Unsupported system: ${system}");
+in
+  stdenvNoCC.mkDerivation rec {
+    inherit
+      version
+      ;
```
Comment (pbsds) [inline pkgs/development/tools/cosmo/default.nix:23]:
```suggestion
    inherit version;
```

**Pull request closed without merging**